lib/stages/ReadGain.cpp

```diff
@@ -0,0 +1,230 @@
+using kotekan::Config;
+using kotekan::configUpdater;
+using kotekan::Stage;
+
+using kotekan::connectionInstance;
+using kotekan::HTTP_RESPONSE;
+using kotekan::restServer;
+
+REGISTER_KOTEKAN_STAGE(ReadGain);
+
+// clang-format off
+
+// Request gain file re-parse with e.g.
+// FRB
+// curl localhost:12048/frb_gain -X POST -H 'Content-Type: appication/json' -d '{"frb_gain_dir":"the_new_path"}'
+// PSR
+// curl localhost:12048/updatable_config/pulsar_gain -X POST -H 'Content-Type: application/json' -d
+// '{"pulsar_gain_dir":["path0","path1","path2","path3","path4","path5","path6","path7","path8","path9"]}'
+//
+// clang-format on
+
+ReadGain::ReadGain(Config& config, const std::string& unique_name,
+                   bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&ReadGain::main_thread, this)) {
+    // Apply config.
+    _num_elements = config.get<uint32_t>(unique_name, "num_elements");
+    _num_beams = config.get<int32_t>(unique_name, "num_beams");
+    scaling = config.get_default<float>(unique_name, "frb_scaling", 1.0);
+    vector<float> dg = {0.0, 0.0}; // re,im
+    default_gains = config.get_default<std::vector<float>>(unique_name, "frb_missing_gains", dg);
+
+    metadata_buf = get_buffer("in_buf");
+    register_consumer(metadata_buf, unique_name.c_str());
+    metadata_buffer_id = 0;
+    metadata_buffer_precondition_id = 0;
+    freq_idx = -1;
+    freq_MHz = -1;
+
+    // Gain for FRB
+    gain_frb_buf = get_buffer("gain_frb_buf");
+    gain_frb_buf_id = 0;
+    register_producer(gain_frb_buf, unique_name.c_str());
+    update_gains_frb = true;
+
+    // Gain for PSR
+    gain_psr_buf = get_buffer("gain_psr_buf");
+    gain_psr_buf_id = 0;
+    register_producer(gain_psr_buf, unique_name.c_str());
+    update_gains_psr = true;
+
+
+    using namespace std::placeholders;
+
+    // listen for gain updates FRB
+    _gain_dir_frb = config.get_default<std::string>(unique_name, "updatable_config/gain_frb", "");
+    if (_gain_dir_frb.length() > 0)
+        configUpdater::instance().subscribe(
+            config.get<std::string>(unique_name, "updatable_config/gain_frb"),
+            std::bind(&ReadGain::update_gains_frb_callback, this, _1));
+
+    // listen for gain updates PSR
+    configUpdater::instance().subscribe(
+        config.get<std::string>(unique_name, "updatable_config/gain_psr"),
+        std::bind(&ReadGain::update_gains_psr_callback, this, _1));
+}
+
+bool ReadGain::update_gains_frb_callback(nlohmann::json& json) {
+    {
+        std::lock_guard<std::mutex> lock(mux);
+        update_gains_frb = true;
+    }
+    cond_var.notify_all();
+
+    try {
+        _gain_dir_frb = json.at("frb_gain_dir");
+    } catch (std::exception& e) {
+        WARN("[FRB] Fail to read gain_dir {:s}", e.what());
+        return false;
+    }
+    return true;
+}
+
+bool ReadGain::update_gains_psr_callback(nlohmann::json& json) {
+    {
+        std::lock_guard<std::mutex> lock(mux);
```

Comment:
Same problem here as above.    And actually there is another problem here.  If a gain update is called while `update_gains_psr == true` then it's possible you'll never update the gains with the latest file since in the main thread you'll set `update_gains_psr = false` before looping back.    There are three ways to fix this: 1) check if `update_gains_psr == true` and if so reject the update with a warning you cannot handle two updates at once 2) lock the entire `read_gain_psr(); update_gains_psr = false;` sections, or 3) have an update queue.   I think 1) might be the best bet for now.  I don't think we'll ever want to send updates that fast, and it's better than 2) which locks up the REST server for a while, and 3) is a bit more work...